types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,630 @@
+	struct wlr_drm_lease_device_v1 *device =
+		wl_container_of(listener, device, backend_destroy);
+
+	wlr_drm_lease_device_v1_destroy(device);
+}
+
+static struct wlr_drm_lease_device_v1 *wlr_drm_lease_device_v1_create(
+		struct wl_display *display, struct wlr_backend *backend) {
+	assert(display && backend && wlr_backend_is_drm(backend));
+
+	struct wlr_drm_backend *drm_backend =
+			get_drm_backend_from_backend(backend);
+	wlr_log(WLR_DEBUG, "Creating wlr_drm_lease_device_v1 for %s",
+			drm_backend->name);
+
+	struct wlr_drm_lease_device_v1 *lease_device =
+		calloc(1, sizeof(struct wlr_drm_lease_device_v1));
+
+	if (!lease_device) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_device_v1");
+		return NULL;
+	}
+
+	lease_device->backend = backend;
+	wl_list_init(&lease_device->resources);
+	wl_list_init(&lease_device->connectors);
+	wl_list_init(&lease_device->requests);
+	wl_list_init(&lease_device->leases);
+	wl_list_init(&lease_device->link);
+
+	lease_device->backend_destroy.notify = handle_backend_destroy;
+
+	wl_signal_add(&backend->events.destroy, &lease_device->backend_destroy);
+
+	lease_device->global = wl_global_create(display,
+		&wp_drm_lease_device_v1_interface, DRM_LEASE_DEVICE_V1_VERSION,
+		lease_device, lease_device_bind);
+
+	if (!lease_device->global) {
+		free(lease_device);
+		return NULL;
+	}
+
+	return lease_device;
+}
+
+struct multi_backend_data {
+	struct wl_display *display;
+	struct wlr_drm_lease_manager *manager;
+};
+
+static void multi_backend_cb(struct wlr_backend *backend, void *data) {
+	struct multi_backend_data *backend_data = data;
+	if (!wlr_backend_is_drm(backend)) {
+		return;
+	}
+
+	struct wlr_drm_lease_device_v1 *device = wlr_drm_lease_device_v1_create(
+			backend_data->display, backend);
+
+	device->manager = backend_data->manager;
+
+	wl_list_insert(&backend_data->manager->devices, &device->link);
+}
+
+static void handle_display_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_drm_lease_manager *manager = wl_container_of(listener, manager,
+			display_destroy);
+
+	wlr_log(WLR_DEBUG, "Destroying wlr_drm_lease_manager");
+
+	struct wlr_drm_lease_device_v1 *device, *tmp;
+	wl_list_for_each_safe(device, tmp, &manager->devices, link) {
+		wl_list_remove(&device->link);
+		wl_list_init(&device->link);
+		wlr_drm_lease_device_v1_destroy(device);
+	}
+}
+
+struct wlr_drm_lease_manager *wlr_drm_lease_manager_create(
+		struct wl_display *display, struct wlr_backend *backend) {
+	struct wlr_drm_lease_manager *manager = calloc(1,
+			sizeof(struct wlr_drm_lease_manager));
+	if (!manager) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_manager");
+		return NULL;
+	}
+
+	wl_list_init(&manager->devices);
+
+	if (wlr_backend_is_multi(backend)) {
+		struct multi_backend_data data = {
+			.display = display,
+			.manager = manager,
+		};
+		wlr_multi_for_each_backend(backend, multi_backend_cb, &data);
```

Comment:
We should probably fail if the multi backend didn't contain any DRM backend here.